### ISSUE TITLE
Item-Caption broken in dev master

### ISSUE THREAD

**Issue by aitte2** [mgp25/Instagram-API#1612]
Hi everyone! :)

This commit has just broken captions:

https://github.com/mgp25/Instagram-API/commit/d62b959906995a7aee0f82d1fafcf61dc3b83f58

The old code was correct. Captions _are_ meant to be this object:
https://github.com/mgp25/Instagram-API/blob/master/src/Response/Model/Caption.php

Here is a caption I captured now, and its contents perfectly match the Caption object you had before:

```
                "caption": {
                    "pk": "17876224726178144",
                    "user_id": "205593849",
                    "text": "2017 Rushing Yards Leaders! (Through Week 5)",
                    "type": 1,
                    "created_at": 1507676114,
                    "created_at_utc": 1507676114,
                    "content_type": "comment",
                    "status": "Active",
                    "bit_flags": 0,
                    "user": {
                        "pk": "205593849",
                        "username": "nfl",
                        "full_name": "NFL",
                        "is_private": false,
                        "profile_pic_url": "https:\/\/scontent-arn2-1.cdninstagram.com\/t51.2885-19\/s150x150\/22277531_151132242155064_7627195811218063360_n.jpg",
                        "profile_pic_id": "1621140123620993470_205593849",
                        "friendship_status": {
                            "following": true,
                            "outgoing_request": false,
                            "is_bestie": false
                        },
                        "is_verified": true,
                        "has_anonymous_profile_picture": false,
                        "is_unpublished": false,
                        "is_favorite": false
                    },
                    "did_report_as_spam": false,
                    "media_id": "1622827317217699596"
                },
```

Anyone who is reading captions should not upgrade dev-master. The latest commits made it impossible to read captions, until the Caption object is restored.

I had to downgrade my instagram-api. Was this change an accident? :)

I noticed the problem immediately after upgrade because my script crashed because caption could no longer be read.

Here is some proof to show that Caption object is needed!

```php
$res = $ig->timeline->getTimelineFeed()->getFeedItems()[0]->getMediaOrAd()->getCaption();
var_dump($res);
```

Stuff like this command is now broken:

```php
$user = $ig->timeline->getTimelineFeed()->getFeedItems()[0]->getMediaOrAd()->getCaption()->getUser();
```

Please restore the Caption object.

PS: I have never used story attribution so I don't know, but those were removed at the same time. Is that change a bug too? https://github.com/mgp25/Instagram-API/commit/bd262fe526b8ba8a8dcecef64679c098a6b12bfc

**Comment by aitte2** [comment]
PS: I have never seen story attribution so I don't know, but you've removed those too. Is that change a bug too? https://github.com/mgp25/Instagram-API/commit/bd262fe526b8ba8a8dcecef64679c098a6b12bfc

**Comment by mgp25** [comment]
Hello @aitte2, 

that commit was an issue, ill restore it. Thanks for reporting in. 

I dont recommend you using dev-master at the moment, because there are still things that doesnt match the current version. If you could use the latest release it would be better.

Regards

**Comment by mgp25** [comment]
Hello @aitte2,

It should be fixed in this commit: https://github.com/mgp25/Instagram-API/commit/81cd09fc6cb32e93f7cb3c86efb3b1ffe54b101d

Regards

**Comment by aitte2** [comment]
@mgp25 Thanks. But since you didn't re-add it to the "property map", other functions like "isCaption" to check if a caption exists doesn't work anymore. So it's still broken.

I think I am understanding why you did the change now.

It seems like you committed this change together with many Direct response changes. So I guess in Direct messaging, `"caption"` is not an object and couldn't be mapped in the new mapper?

I think I have a solution.

Please try something like this:

`src/Response/Model/Item.php` (Make this one as usual)

```php
class Item extends AutoPropertyMapper
{
    const JSON_PROPERTY_MAP = [
        ...
        'caption' => 'Caption',
    ];
}
```

`src/Response/Model/Direct/Item.php` (Make this one extend regular Item, but override what "caption" maps into)

```php
use InstagramAPI\Response\Model;

class Item extends Model\Item
{
    const JSON_PROPERTY_MAP = [
        'caption' => 'string',
    ];
}
```

Finally, just make `Direct`-related responses/models map into `Model\Direct\Item` instead of `Model\Item`.

If I understand everything correctly, this will mean that Direct items use all usual Item properties _plus_ override "caption" to be a string instead.

Please try it. It seems to work for me. I tried it manually and the final "Model\Direct\Item" map has a string caption and a regular "Model\Item" has a Caption caption. :) 👍

**Comment by mgp25** [comment]
@aitte2 yes! that is exactly what happened! I'll check it and finally fix this.

Thanks, i'll let you know when i push the fix.

Thanks again

Regards

**Comment by aitte2** [comment]
I forgot about StoryLocation (with its Attribution stuff).

`src/Response/Model/Item.php`:

```php
class Item extends AutoPropertyMapper
{
    const JSON_PROPERTY_MAP = [
        ...
        'caption'         => 'Caption',
        'story_locations' => 'StoryLocation[]',
    ];
}
```

`src/Response/Model/Direct/DirectStoryLocation.php`

```php
namespace InstagramAPI\Response\Model\Direct;

use InstagramAPI\Response\Model;

class DirectStoryLocation extends Model\StoryLocation
{
    const JSON_PROPERTY_MAP = [
        'attribution' => '', // What type?
    ];
}
```

`src/Response/Model/Direct/DirectItem.php`

```php
namespace InstagramAPI\Response\Model\Direct;

use InstagramAPI\Response\Model;

class DirectItem extends Model\Item
{
    const JSON_PROPERTY_MAP = [
        'caption'         => 'string',
        'story_locations' => 'DirectStoryLocation[]',
    ];
}
```

And finally edit `src/Response/Model/DirectThreadItem.php`, to point `media_share` and `raven_media` at `Direct\DirectItem`.

PS: I can make this as a pull request.

**Comment by aitte2** [comment]
@mgp25 In case you are working on this right now, please reload this page and look at my final message above, I edited it to suggest a better name for the new classes and to include the StoryLocation "attribution" problem too. I think I have figured out a full fix now. Please let me know if this goes well :)

**Comment by mgp25** [comment]
Hello @aitte2, im finishing some homework at the moment 😄 but as soon as i finish them, ill check all your comments and see your fix.

Regards

**Comment by aitte2** [comment]
@mgp25 Great, good luck ❤️  :)

**Comment by mgp25** [comment]
Hello @aitte2, sorry for the delay, can you send a PR?

**Comment by aitte2** [comment]
Hi @mgp25, do you really need me to send a PR? :D Every step is described in a single post here:

https://github.com/mgp25/Instagram-API/issues/1612#issuecomment-335816442

I don't have any Direct thread to test with to be 100% sure, so I'd rather let you do the changes since you are able to test it afterwards. Just do the steps I mentioned there. :)

**Comment by mgp25** [comment]
Hello @aitte2, i tested your suggestion, but it gives me an error. Let me show you the response:

`GET:  https://i.instagram.com/api/v1/news/inbox/?activity_module=all&show_su=true
→ 0B
← 200 	 5.99kB
RESPONSE: {"aymf": {"items": [{"user": {"pk": "uid", "username": "someUsername", "full_name": "someUsername", "is_private": true, "profile_pic_url": "https://scontent-mad1-1.cdninstagram.com/t51.2885-19/s150x150/ANOTHERID_n.jpg", "profile_pic_id": "MEDIA_ID", "is_verified": false, "has_anonymous_profile_picture": false}, "algorithm": "contact_importer_algorithm", "social_context": "In your contacts", "icon": "", "caption": "", "media_ids": [], "thumbnail_urls": [], "large_urls": [], "media_infos": [],
. . .`


`Fatal error: Uncaught InstagramAPI\Exception\InstagramException: Bad JSON property definitions in "InstagramAPI\Response\Model\Item": (['Unable to convert non-array inner value for property "caption" into class "\InstagramAPI\Response\Model\Caption".']). in /Users/mgp25/Desktop/Instagram-API/src/Client.php:563`

**Comment by aitte2** [comment]
@mgp25 Okay. The problem above shows that the JSON data is a mismatch in multiple endpoints. This is a problem of re-using `Item` for everything and thinking it's always gonna be the same. It's not. So I guess it's time to start clearing that up in this project? :P

I can see that `news/inbox` is from `$ig->people->getRecentActivityInbox()` which is an ActivityNewsResponse. And it has an `aymf` property which links to `Model\Aymf.php` which has an "items" array with `Item[]`.

That's the problem! Aymf items are not the same as normal `Item`.

This needs to be solved with more classes:

`src/Response/Model/AymfItem.php`:

```php
class AymfItem extends Item
{
    const JSON_PROPERTY_MAP = [
        'caption'         => 'string',
    ];
}
```

And then just update `src/Response/Model/Aymf.php` to use `AymfItem` instead of `Item`.

Any time you see more problems like this, use this solution. :)

**Comment by aitte2** [comment]
I will not use dev-master anymore until this is safe hehe, seems like risky stuff until all bad old classes are mappable.

**Comment by mgp25** [comment]
@aitte2 do i keep previous changes too? Well, ill read the comment probably when i wake up.

And yes, we should do a clean up, because there are unused params, outdated params, and maybe duplicate object

**Comment by aitte2** [comment]
@mgp25 The previous changes may not be good. I thought DirectThreadItem was the problem. Revert those until you know for sure what DirectThreadItem needs as its caption/story attribution types.

I see that DirectThreadItem is used by DirectVisualThreadResponse.php, so get such a response and look at its item captions and see if they are Caption objects (like normal Items are) or if they are strings (like AymfItems are). Because the changes I suggested earlier assumed that Direct threads were the ones that needed these fixes.

About the clean up: Yeah I guess so but it sounds so hard, are you sure you want to? There's so many properties and so many responses that reuse them that you would have to touch every API and try them all with lots of data to find all data that is truly not needed. It doesn't seem worth it. I would just let the old properties exist without caring about them :)

**Comment by aitte2** [comment]
@mgp25 I will do a pull request with the things I find. I will take care of AymfItem.

**Comment by aitte2** [comment]
@mgp25 https://github.com/mgp25/Instagram-API/pull/1614

:)

This was how I found what to do:

In your log (and in my tests), we can see what endpoint gave us JSON before the crash happened.

So I saw ` https://i.instagram.com/api/v1/news/inbox/?activity_module=all&show_su=true`

Then I searched in src/Request/ for `news/inbox` and found `$ig->people->getRecentActivityInbox()`.

Then I looked at its response type which is `ActivityNewsResponse`.

Then I looked at all properties inside `ActivityNewsResponse` to find what was linking us to Item.

It was `Aymf`, which linked to `Item`.

So I made a new one called `AymfItem`, which extends `Item` but overrides `"caption"` to just be a string instead of a Caption object.

Finally, I edited Aymf to use the new `AymfItem` instead.

And yup it works! :)
